js/ui/scheduler/ui.scheduler.recurrence_editor.js

```diff
@@ -199,373 +195,311 @@ const RecurrenceEditor = Editor.inherit({
+                            },
+                            onValueChanged: (args) => this._valueChangedHandler(args)
+                        },
+                        label: {
+                            text: messageLocalization.format('dxScheduler-recurrenceRepeatEvery')
+                        }
+                    },
+                    {
+                        name: 'intervalLabel',
+                        cssClass: `${INTERVAL_EDITOR}${LABEL_POSTFIX}`,
+                        template: () => messageLocalization.format(`dxScheduler-recurrenceRepeat${freq.charAt(0).toUpperCase()}${freq.substr(1).toLowerCase()}`)
+                    }
+                ]
+            },
+            {
+                itemType: 'group',
+                items: [
+                    {
+                        name: 'repeatOnLabel',
+                        colSpan: 2,
+                        template: () => messageLocalization.format('dxScheduler-recurrenceRepeatOn'),
+                        visible: freq && freq !== 'daily' && freq !== 'hourly',
+                    }
+                ]
+            },
+            {
+                itemType: 'group',
+                cssClass: REPEAT_ON_EDITOR,
+                colCount: 2,
+                colCountByScreen: { xs: 2 },
+                items: this._getRepeatOnItems(freq)
+            },
+            {
+                itemType: 'group',
+                items: this._getRepeatEndItems()
             }
-        });
+        ];
 
-        const $field = $('<div>')
-            .addClass(FIELD_CLASS)
-            .addClass(RECURRENCE_FREQ_FIELD)
-            .append($freqEditorLabel, $freqEditor);
-
-        this._$container.append($field);
+        return this._editors;
     },
 
-    _renderIntervalEditor() {
-        const freq = this._recurrenceRule.rules().freq || 'daily';
-
-        const $intervalEditor = $('<div>')
-            .addClass(INTERVAL_EDITOR)
-            .addClass(FIELD_VALUE_CLASS);
-
-        const $intervalEditorLabel = $('<div>')
-            .text(messageLocalization.format('dxScheduler-recurrenceRepeatEvery'))
-            .addClass(INTERVAL_EDITOR + LABEL_POSTFIX)
-            .addClass(FIELD_LABEL_CLASS);
+    _getRepeatOnItems: function(freq) {
+        const monthsNames = dateLocalization.getMonthNames('wide');
+        const months = [];
 
-        this._$intervalTypeLabel = $('<div>')
-            .text(messageLocalization.format(`dxScheduler-recurrenceRepeat${freq.charAt(0).toUpperCase()}${freq.substr(1).toLowerCase()}`))
-            .addClass(REPEAT_TYPE_EDITOR + LABEL_POSTFIX);
+        for(let i = 0; i < 12; i++) {
+            months[i] = { value: String(i + 1), text: monthsNames[i] };
+        }
 
-        const interval = this._recurrenceRule.rules().interval || 1;
+        return [
+            {
+                dataField: 'byday',
+                colSpan: 2,
+                template: (_, itemElement) =>{
```

Comment:
To my mind `Template` should be in a separate method.

Reply:
Yes, it would be wise to fix it